### PR TITLE
Add shutdown and safe_mode button

### PR DESCRIPTION
**Pull request by jsuanet**
# What does this implement/fix? 

New buttons for shutdown and safe_mode, can be used as replacement for shutdown and safe_mode switches.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#1725

## Test Environment

- [X] ESP32
- [ ] ESP32 IDF
- [X] ESP8266

## Example entry for `config.yaml`:

```yaml
button:
  - platform: safe_mode
    name: my_device safe_mode button
  - platform: shutdown
    name: my_device shutdown button

```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: esphome/components/safe_mode/button/__init__.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+SafeModeButton = safe_mode_ns.class_("SafeModeButton", button.Button, cg.Component)
+
+CONFIG_SCHEMA = (
+    button.button_schema(
+        device_class=DEVICE_CLASS_RESTART, entity_category=ENTITY_CATEGORY_CONFIG
+    )
+    .extend({cv.GenerateID(): cv.declare_id(SafeModeButton)})
+    .extend({cv.GenerateID(CONF_OTA): cv.use_id(OTAComponent)})
+    .extend({cv.Optional(CONF_ICON, default=ICON_RESTART_ALERT): cv.icon})
+    .extend(cv.COMPONENT_SCHEMA)
```
Comment (jesserockz) [inline esphome/components/safe_mode/button/__init__.py]: The icon can be set in the `button_schema()`

File: esphome/components/safe_mode/button/safe_mode_button.cpp [hunk 2]
```diff
@@ -0,0 +1,25 @@
+#include "safe_mode_button.h"
+#include "esphome/core/hal.h"
+#include "esphome/core/log.h"
+#include "esphome/core/application.h"
+
+namespace esphome {
+namespace safe_mode {
+
+static const char *const TAG = "safe_mode_button";
```
Comment (jesserockz) [inline esphome/components/safe_mode/button/safe_mode_button.cpp]: Tags should always be separated with a `.`

```suggestion
static const char *const TAG = "safe_mode.button";
```

File: esphome/components/shutdown/button/__init__.py [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+ShutdownButton = shutdown_ns.class_("ShutdownButton", button.Button, cg.Component)
+
+CONFIG_SCHEMA = (
+    button.button_schema(
+        device_class=DEVICE_CLASS_RESTART, entity_category=ENTITY_CATEGORY_CONFIG
+    )
+    .extend({cv.GenerateID(): cv.declare_id(ShutdownButton)})
+    .extend({cv.Optional(CONF_ICON, default=ICON_POWER): cv.icon})
+    .extend(cv.COMPONENT_SCHEMA)
+)
```
Comment (jesserockz) [inline esphome/components/shutdown/button/__init__.py]: Move icon into `button_schema` and remove device_class as it does not match.

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/safe_mode/button/__init__.py (hunk 1 above)
Comment (jsuanet) [reply esphome/components/safe_mode/button/__init__.py]: done

File: esphome/components/shutdown/button/__init__.py (hunk 3 above)
Comment (jsuanet) [reply esphome/components/shutdown/button/__init__.py]: done

**Review by jesserockz — APPROVED** [review]
LGTM

**Review by oxan — APPROVED** [review]
At some point we can also deprecate the old switches.

**Comment by oxan** [comment]
Thanks!

**Pull request merged**